yktrace/src/tir.rs

```diff
@@ -41,6 +41,11 @@ impl<'a, 'm> TirTrace<'a, 'm> {
         // Maps symbol names to their virtual addresses.
         let mut addr_map: HashMap<String, u64> = HashMap::new();
 
+        // We use a stack of IPlaces to keep track of storage for return values of inlined calls.
+        // When we encounter the Return terminator of a call that's being inlined into the trace,
```

Comment:
This sentence is very hard to parse (I eventually worked out it out, but it took me a few reads). I'd suggest starting with what it *does* not what it *doesn't do*. Something like "We need to convert `return`s from inlined functions into stores" maybe?

Reply:
Try 968e2a8, is this clearer?

Reply:
I'm not sure what "To avoid unnecessary stores" has to do with anything?

Reply:
This is effectively an optimisation that eliminates one store.

Instead of:
```
callee:
  $x = return value
caller:
  $y = $x
```

You get:
```
callee:
   $y = return value
```

Reply:
What Edd said. This is an optimisation that assigns the return value of an inlined function immediately to the destination of the function call instead of going via the return local (`$0`).

Reply:
Ah, then I misunderstood what's going on. Silly question: why are we doing this optimisation here? Wouldn't it be better to leave this to a generic optimisation pass?

Reply:
I would guess as it is very common and is much cheaper to do in this case than the general case.

Reply:
I'm asking partly for a generic reason: the problem with "early" optimisations like these is that they tend to introduce bugs (like the one cleared up by this PR). Then we have to ask whether they're worth it, or, at least, whether they're worth doing now. I'm open-minded in this case!

Reply:
Since it's implemented, I'd say leave it for now.

Reply:
Are we reasonably confident that a) there aren't other edge cases or b) that we've thought through those edge cases?

Reply:
At the time (when we didn't have `do_not_trace` yet) it was a very simple optimisation which essentially did this:
```
$0 = $1;
$5 = $0;
```
becomes
```
$5 = $1;
```
So there wasn't too much to think about in terms of edge cases.